spicy/toolchain/src/compiler/validator.cc

```diff
@@ -623,9 +623,16 @@ struct VisitorPost : visitor::PreOrder, hilti::validator::VisitorMixIn {
         else if ( n->tag() == "&requires" ) {
             if ( ! n->hasValue() )
                 error("&requires must provide an expression", n);
-            else if ( auto e = n->valueAsExpression();
-                      e && ! hilti::type::same((*e)->type()->type(), builder.typeBool()) )
-                error(fmt("&requires expression must be of type bool, but is of type %d ", *(*e)->type()), n);
+            else {
+                auto e = *n->valueAsExpression();
+                assert(e);
+
+                if ( auto result = e->type()->type()->tryAs<hilti::type::Result>();
+                     ! result || ! result->dereferencedType()->type()->isA<hilti::type::Void>() )
+                    error(fmt("&requires expression must be of type bool or result<void>, but is of type %d ",
+                              *e->type()),
```

Comment:
Nit: we don't really care about the value, but its stringification.

```suggestion
                              e->type()->print()),
```

Reply:
The `*`-style is how it's done almost everywhere (mostly for historical reasons) so staying with that.